ckanext/switzerland/logic.py

```diff
@@ -157,3 +163,106 @@ def highlight(term, q):
+        sources_to_clear = [source.id]
+    else:
+        all_sources = model.Session.query(HarvestSource.id).all()
+        sources_to_clear = list(itertools.chain(*all_sources))
+
+    log.info('Got sources to clear: {}'.format(','.join(sources_to_clear)))
+
+    # get number of jobs to keep form data_dict
+    if 'number_of_jobs_to_keep' in data_dict:
+        number_of_jobs_to_keep = data_dict['number_of_jobs_to_keep']
+    else:
+        log.error('Configuration missing for number of jobs to keep when clearing harvest sources')
+        raise ValidationError('Configuration missing for number of jobs to keep when clearing harvest sources')
+
+    log.info('Got config for number of most recent jobs per source to keep: {}'.format(number_of_jobs_to_keep))
+
+    # init list of jobs and objects to delete
+    delete_jobs_ids_all = []
+    delete_objects_ids_all = []
+
+    for source in sources_to_clear:
+        # get jobs ordered by their creations date
+        delete_jobs = model.Session.query(HarvestJob) \
+            .filter(HarvestJob.source_id == source) \
+            .order_by(HarvestJob.created.desc()).all()
+
+        # decide which jobs to keep or delete on their order
+        delete_jobs_ids = [job.id for job in delete_jobs][number_of_jobs_to_keep:]
+        keep_jobs_ids = [job.id for job in delete_jobs][:number_of_jobs_to_keep]
+
+        # log all job for a source with the decision to delete or keep them
+        log.info('Clear harvest source {}'.format(source))
+        log.info('- jobs_to_keep: {}'.format(','.join(keep_jobs_ids)))
+        for job in delete_jobs[:number_of_jobs_to_keep]:
+            log.debug('    - job {}: created:{}, status:{}'.format(job.id, job.created, job.status))
+        log.info('- jobs to delete: {}'.format(','.join(delete_jobs_ids)))
+        for job in delete_jobs[number_of_jobs_to_keep:]:
+            log.debug('    - job {}: created:{}, status:{}'.format(job.id, job.created, job.status))
+
+        # add to list of jobs to delete
+        delete_jobs_ids_all.extend(delete_jobs_ids)
+
+    # log all jobs to delete
+    log.info('Jobs to delete in total {}'.format(delete_jobs_ids_all))
+
+    if delete_jobs_ids_all:
+        # get harvest objects for harvest jobs
+        delete_objects = model.Session.query(HarvestObject.id) \
+           .filter(HarvestObject.harvest_job_id.in_(delete_jobs_ids_all)).all()
+        delete_objects_ids_all = list(itertools.chain(*delete_objects))
+
+        # perform delete
+        sql = '''begin;
+        delete from harvest_object_error where harvest_object_id in ('{delete_objects_values}');
+        delete from harvest_object_extra where harvest_object_id in ('{delete_objects_values}');
+        delete from harvest_object where id in ('{delete_objects_values}');
+        delete from harvest_gather_error where harvest_job_id in ('{delete_jobs_values}');
+        delete from harvest_job where id in ('{delete_jobs_values}');
```

Comment:
What does the job do, if it fails executing the SQL-command? As there are quite a few constraints at play could you look into what exceptions might be thrown and except them to desired behaviour? Maybe it makes sense in that case to move on to the next source? 
Not sure if this way we could end up with old 'broken' log entries that can never be deleted but delete the newer ones. What do you think?

Reply:
I compared all the places in OGDZH and OGDCH where sql is executed like this. The result is never evaluated and I looked into the result and was unsure what could be seen there. I copied this from the harvester which proceeds exactly this way. Therefore I concluded this would be good enough like this.